### PR TITLE
add a single sirun test

### PR DESCRIPTION
**Pull request by bengl**
### What does this PR do?
As an initial prototype, add a sirun test.
### Motivation
We want to move forward with using sirun for benchmarking.

### CODE REVIEW

File: benchmark/sirun/startup/meta.json [hunk 1]
```diff
@@ -0,0 +1,13 @@
+{
+  "name": "startup",
+  "run": "node startup-test.js",
```
Comment (brettlangdon) [inline benchmark/sirun/startup/meta.json:3]: Does it matter if we use the same script or not?

```json
{
  "name": "startup",
  "cachegrind": true,
  "variants": {
    "control": { "run": "node -e ''" },
    "with-tracer": { "run": "node -e 'require('../../../').init()" }
  }
}
```


It actually doesn't matter at all, one way or the other, just curious if there is a reason to choose one approach over the other (mostly for my own knowledge when I write some scenarios)

**Review by brettlangdon — DISMISSED** [review]

**Review by rochdev — CHANGES_REQUESTED** [review]
This is not runnable locally in its current state. Also, I thought sirun was in charge of running the variants which doesn't seem to be the case.

File: benchmark/sirun/startup/meta.json (hunk 1 above)
Comment (bengl) [reply benchmark/sirun/startup/meta.json:3]: It doesn't matter. This was purely personal preference. I did want to use a file though, rather than `-e` to be at least a *little* closer to what users actually do. The `-e` environment isn't _quite_ the same as running a script.

Comment (brettlangdon) [reply benchmark/sirun/startup/meta.json:3]: makes sense

**Review by rochdev — DISMISSED** [review]

File: .circleci/config.yml [hunk 2]
```diff
@@ -69,6 +69,31 @@ node-bench-e2e-base: &node-bench-e2e-base
...
+        name: Install system deps
+        command: apt-get update && apt-get install -y valgrind wget
+    - run:
+        name: Install sirun
+        command: wget https://github.com/DataDog/sirun/releases/download/v0.1.3/sirun-v0.1.3-x86_64-unknown-linux-gnu.tar.gz && tar zxf sirun-v0.1.3-x86_64-unknown-linux-gnu.tar.gz && mv sirun /usr/bin
+    - run:
+        name: Run sirun
+        command: |
+          cd $SIRUN_TEST_DIR
+          node /root/dd-trace-js/benchmark/sirun/run-all-variants.js | tee /tmp/sirun-output.ndjson
```
Comment (brettlangdon) [inline .circleci/config.yml:91]: will this also write any stdout from the script to the file?

Comment (bengl) [reply .circleci/config.yml:91]: stdout and stderr from tested programs are suppressed in this PR with `SIRUN_NO_STDIO=1`.

Comment (brettlangdon) [reply .circleci/config.yml:91]: ah, thanks, I knew I was missing something

File: benchmark/sirun/run-all-variants.js [hunk 3]
```diff
@@ -0,0 +1,18 @@
+'use strict'
```
Comment (rochdev) [inline benchmark/sirun/run-all-variants.js:1]: Is there a plan to move all this logic to sirun since it will need to be duplicated in all languages?

File: benchmark/sirun/get-results.js [hunk 4]
```diff
@@ -0,0 +1,73 @@
+'use strict'
```
Comment (rochdev) [inline benchmark/sirun/get-results.js:1]: What is the purpose of this file? Is it something sirun should do?

File: .circleci/config.yml [hunk 5]
```diff
@@ -69,6 +69,31 @@ node-bench-e2e-base: &node-bench-e2e-base
...
+    - run:
+        name: Install system deps
+        command: apt-get update && apt-get install -y valgrind wget
+    - run:
+        name: Install sirun
+        command: wget https://github.com/DataDog/sirun/releases/download/v0.1.3/sirun-v0.1.3-x86_64-unknown-linux-gnu.tar.gz && tar zxf sirun-v0.1.3-x86_64-unknown-linux-gnu.tar.gz && mv sirun /usr/bin
+    - run:
+        name: Run sirun
+        command: |
+          cd $SIRUN_TEST_DIR
```
Comment (rochdev) [inline .circleci/config.yml:90]: Why is this needed?

File: benchmark/sirun/startup/meta.json [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+  "cachegrind": true,
+  "iterations": 10,
+  "variants": {
+    "control": {},
```
[the remark this answers is not in the document]
Comment (bengl) [reply benchmark/sirun/startup/meta.json:7]: Sure.

File: benchmark/sirun/startup/meta.json [hunk 7]
```diff
@@ -0,0 +1,14 @@
+{
+  "name": "startup",
+  "run": "node startup-test.js",
+  "cachegrind": true,
+  "iterations": 10,
```
[the remark this answers is not in the document]
Comment (bengl) [reply benchmark/sirun/startup/meta.json:5]: This runs the test 10 times (and an additional time for cachegrind). Why? That way we can see if there's any variance here.

File: .circleci/config.yml (hunk 5 above)
Comment (bengl) [reply .circleci/config.yml:90]: This is how we're selecting which test(s) to run. The variable is set in the specific case defined later in the file.

Comment (rochdev) [reply .circleci/config.yml:90]: Could this be done without having to `cd` into each folder? Does sirun support specifying any path to a `meta.json`?

File: benchmark/sirun/startup/meta.json (hunk 7 above)
[the remark this answers is not in the document]
Comment (rochdev) [reply benchmark/sirun/startup/meta.json:5]: I would argue that if there is a significant variance then running it multiple times won't really help increase confidence. Does this capture the data as a histogram?

File: benchmark/sirun/get-results.js (hunk 4 above)
Comment (rochdev) [reply benchmark/sirun/get-results.js:1]: Wasn't there supposed to be support for different target environments like CircleCI to generate the YAML etc? Also, I still don't understand what is the purpose of this file.

File: benchmark/sirun/run-all-variants.js (hunk 3 above)
Comment (bengl) [reply benchmark/sirun/run-all-variants.js:1]: Because for some folks/use-cases it might make sense to tear down an environment completely between variants (e.g. start a new container, etc.), it seemed to make more sense to just have sirun run one variant at a time, and leave any orchestration of that up to users.

File: benchmark/sirun/startup/meta.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (bengl) [reply benchmark/sirun/startup/meta.json]: Fixed in latest push.

File: .circleci/config.yml (hunk 5 above)
Comment (bengl) [reply .circleci/config.yml:90]: > Could this be done without having to cd into each folder?

Sure, but you'll want the `$PWD` to be this directory anyway, so that the commands in the meta.json can use relative paths.

> Does sirun support specifying any path to a meta.json?

Not at the moment. I'm not sure how that's beneficial over just setting the working directory here though.

File: benchmark/sirun/get-results.js (hunk 4 above)
Comment (bengl) [reply benchmark/sirun/get-results.js:1]: > Wasn't there supposed to be support for different target environments like CircleCI to generate the YAML etc?

Yeah. But that was only really discussed in terms of defining tests, not grabbing their results.

> Also, I still don't understand what is the purpose of this file.

The file exists as a quick-and-easy tool to get all the sirun test results for a given commit from CircleCI artifacts. If that's objectionable, I can remove it from the PR.

File: benchmark/sirun/startup/meta.json (hunk 7 above)
[the remark this answers is not in the document]
Comment (bengl) [reply benchmark/sirun/startup/meta.json:5]: > I would argue that if there is a significant variance then running it multiple times won't really help increase confidence. 

Okay, but you won't know about the variance unless you collect the data.

> Does this capture the data as a histogram?

Just raw data. We can do some histograms later on in `get-results.js` or any other analysis tools we create/use.

File: benchmark/sirun/run-all-variants.js (hunk 3 above)
Comment (bengl) [reply benchmark/sirun/run-all-variants.js:1]: On second thought, there's definitely room for at least an _option_ in sirun to run all the variants. I'll make that change in sirun when I have some time. In the meantime, we'll keep this script around. Sound good?

**Review by rochdev — DISMISSED** [review]

File: benchmark/sirun/get-results.js [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+const https = require('https')
+const { execSync } = require('child_process')
+
+const { CIRCLE_TOKEN } = process.env
+
+const ref = process.argv.length > 2 ? process.arv[2] : 'HEAD'
+const gitCommit = execSync(`git rev-parse ${ref}`).toString().trim()
+
+const circleHeaders = {
+  'circle-token': CIRCLE_TOKEN,
```
Comment (brettlangdon) [inline benchmark/sirun/get-results.js:12]: shouldn't need this. since it is a public project you can hit the API no problem.
```suggestion
```

**Review by brettlangdon — APPROVED** [review]

File: benchmark/sirun/get-results.js (hunk 8 above)
Comment (rochdev) [reply benchmark/sirun/get-results.js:12]: I remember having problems when not setting the token, I think what is available and the rate limits are different when not using a token, even for public projects.

Comment (brettlangdon) [reply benchmark/sirun/get-results.js]: I was able to run this script without a token without any issues, but it may be true that rate limits might be an issue.

I have a script I use to pull junit xml results from a bunch of builds (I think it is like 50+ API calls), and I haven't hit any limits or issues yet.

File: benchmark/sirun/get-results.js [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+
+const https = require('https')
+const { execSync } = require('child_process')
+
+const { CIRCLE_TOKEN } = process.env
+
+const ref = process.argv.length > 2 ? process.arv[2] : 'HEAD'
+const gitCommit = execSync(`git rev-parse ${ref}`).toString().trim()
+
+const circleHeaders = CIRCLE_TOKEN ? {
```
Comment (brettlangdon) [inline benchmark/sirun/get-results.js:11]: 👍🏻

**Review by brettlangdon — APPROVED** [review]

File: benchmark/sirun/get-results.js (hunk 8 above)
Comment (bengl) [reply benchmark/sirun/get-results.js]: I changed it so that it will use the token if present, otherwise ignore it.
